### DISCUSSION TITLE
How can miner view startEpoch of deals?

### DISCUSSION THREAD

**Discussion by jburnett** [filecoin-project/lotus#6241] [Storage Provider] [answered]
Clients set startEpoch (24 hrs by default?) on storage deals.  Are miners able to discover this value?

Miners need to see the value due to cases in which startEpoch will occur before miner can complete sealing. If transfer is delayed, for example, the miner needs to be able to determine whether it is worth continuing with the deal.  Once curr epoch + miner's seal time > startEpoch, there's no way for miner to succeed.

**Comment by dd45e640b42e6da7da96faee3996ef7c** [comment, marked as answer]
the proposal the client sends to the miner includes the start epoch - obviously

if you call the external deal filter program - this is what get passed to it:

```
{
    "Proposal": {
      "PieceCID": {
        "/": "baga6ea4sdfsdfsdfsdfsdfsdfnj4qcwwp5jc6gq"
      },
      "PieceSize": 2147483648,
      "VerifiedDeal": false,
      "Client": "f3sdfsdfsdfsdfsdfsdfsdfsdfsdfsdfrpcdmi3urbg36a",
      "Provider": "f099",
      "Label": "Qmsdfsdfsdfsdfsdfsdfsdf9jvKj",
      "StartEpoch": 719986,
      "EndEpoch": 1719986,
      "StoragePricePerEpoch": "46000000000",
      "ProviderCollateral": "567539176515439",
      "ClientCollateral": "0"
    },
    "ClientSignature": {
      "Type": 2,
      "Data": "jB8237sdfsdfsdfsfsdfsdfbZZxl"
    },
    "ProposalCid": {
      "/": "bafddddddddddddddddddddumx3jmohfup754"
    },
    "AddFundsCid": null,
    "PublishCid": null,
    "Miner": "12ffffffffffffffffffffffffffff3",
    "Client": "12bbbbbbbbbbbbbbbbbbbqHYBjw",
    "State": 15,
    "PiecePath": "",
    "MetadataPath": "",
    "SlashEpoch": 0,
    "FastRetrieval": true,
    "Message": "",
    "StoreID": 127,
    "FundsReserved": "0",
    "Ref": {
      "TransferType": "graphsync",
      "Root": {
        "/": "QmyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyjvKj"
      },
      "PieceCid": null,
      "PieceSize": 0,
      "RawBlockSize": 0
    },
    "AvailableForRetrieval": false,
    "DealID": 0,
    "CreationTime": "2021-99-29T34:573:23666.92366666544946Z",
    "TransferChannelId": null,
    "SectorNumber": 0,
    "DealType": "storage"
}
```

**Reply by jburnett** [reply]
Thanks for this answer - just what we're looking for.  What do you mean by "external deal filter program?" A lotus-shed command?

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
>  #Filter = ""
> #RetrievalFilter = ""

excerpt miner config.toml

https://docs.filecoin.io/mine/lotus/miner-configuration/#using-filters-for-fine-grained-storage-and-retrieval-deal-acceptance check this. thats where the data gets passed

**Reply by jburnett** [reply]
Thanks again!

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
if you are in the filecoin slack and have a second - ping me, Patrick -Factor8 Solutions 

i would like to hear what external logic you are looking for and how you plan to implement it
